### ISSUE TITLE
Blue screen on BT.2020 with build dc16d85

### ISSUE THREAD

**Issue by aluo-x** [mpv-player/mpv#5782]
Version: [dc16d85](https://github.com/mpv-player/mpv/commit/dc16d853797d514744998462e3aaa57a6ff7248d)

Platform: Windows 17134.1 64bit. Kaby Lake with driver 24.20.100.6025

Used conf file:

```
vo=gpu
profile=gpu-hq
gpu-context=d3d11
hwdec=d3d11va
tone-mapping=reinhard
#hdr-compute-peak=yes
target-brightness=350
fbo-format=rgb32f
#video-sync=display-resample
interpolation=yes
tscale=linear
audio-channels=stereo
````

Log file: [pastebin link](https://pastebin.com/L5fVZjft)

`fbo-format=rgb32f` is the offending line. Additionally video-sync=display-resample is causing severe stuttering. Was using a March git build before without these issues.

Tested [65f0825](https://github.com/mpv-player/mpv/commit/65f08253158ba9cb4c164a6cfd6da39b39d38e55) to be working with both "display-resample" and rgb32f

**Comment by haasn** [comment]
```
[   0.246][v][vo/gpu] No advanced processing required. Enabling dumb mode.
```

@JEEB seems like the HDR SSBO thing being active doesn't disable dumb mode.

**Comment by haasn** [comment]
Also, maybe we should have the peak detection disabled by default and only enable it in `gpu-hq` mode?

**Comment by jeeb** [comment]
OK, let's compare...

This is my run without any FBO format modification, and yours:
```diff
diff --git a/cygdrive/c/Users/jeeb/my_vo_gpu_log.log b/cygdrive/c/Users/jeeb/users_vo_gpu_log.log
index 5df921d..82525c3 100755
--- a/cygdrive/c/Users/jeeb/my_vo_gpu_log.log
+++ b/cygdrive/c/Users/jeeb/users_vo_gpu_log.log
@@ -1,9 +1,9 @@
 [v][vo/gpu/d3d11] Initializing GPU context 'd3d11'
 [v][vo/gpu/d3d11] Using Direct3D 11 feature level 12_1
-[v][vo/gpu/d3d11] Device Name: NVIDIA GeForce GTX 1080
-[v][vo/gpu/d3d11] Device ID: 10de:1b80 (rev a1)
-[v][vo/gpu/d3d11] Subsystem ID: 1462:3367
-[v][vo/gpu/d3d11] LUID: 000000000000b53d
+[v][vo/gpu/d3d11] Device Name: Intel(R) HD Graphics 630
+[v][vo/gpu/d3d11] Device ID: 8086:591b (rev 04)
+[v][vo/gpu/d3d11] Subsystem ID: 1028:07be
+[v][vo/gpu/d3d11] LUID: 000000000000b760
 [v][vo/gpu/d3d11] Initializing SPIR-V compiler 'shaderc'
 [v][vo/gpu/d3d11] Using Direct3D 11.1 runtime
 [v][vo/gpu/d3d11] Maximum Texture2D size: 16384x16384
@@ -27,7 +27,7 @@
 [d][vo/gpu/d3d11]   rgba16hf   4*float   8B    LF CR {16 16 16 16}
 [d][vo/gpu/d3d11]   r32f       1*float   4B    LF CR {32}
 [d][vo/gpu/d3d11]   rg32f      2*float   8B    LF CR {32 32}
-[d][vo/gpu/d3d11]   rgb32f     3*float  12B          {32 32 32}
+[d][vo/gpu/d3d11]   rgb32f     3*float  12B    LF    {32 32 32}
 [d][vo/gpu/d3d11]   rgba32f    4*float  16B    LF CR {32 32 32 32}
 [d][vo/gpu/d3d11]   rgb10_a2   4*unorm   4B    LF CR {10 10 10 2}
 [d][vo/gpu/d3d11]   bgra8      4*unorm   4B    LF CR {8 8 8 8}
@@ -63,6 +63,7 @@
 [d][vo/gpu/d3d11]   mmal
 [d][vo/gpu/d3d11]   videotoolbox
 [d][vo/gpu/d3d11]   drm_prime
+[d][vo/gpu/d3d11]   cuda
 [d][vo/gpu/d3d11]   yuyv422
 [d][vo/gpu/d3d11]   yuv422p => 3 planes 2x1 8/0 [r8/r8/r8] (r/g/b)
 [d][vo/gpu/d3d11]   yuv410p => 3 planes 4x4 8/0 [r8/r8/r8] (r/g/b)
@@ -162,7 +163,6 @@
 [d][vo/gpu/d3d11]   gbrap16 => 4 planes 1x1 16/0 [r16/r16/r16/r16] (g/b/r/a)
 [d][vo/gpu/d3d11]   qsv
 [d][vo/gpu/d3d11]   d3d11va_vld
-[d][vo/gpu/d3d11]   cuda
 [d][vo/gpu/d3d11]   yuv420p12be
 [d][vo/gpu/d3d11]   yuv420p12 => 3 planes 2x2 16/-4 [r16/r16/r16] (r/g/b)
 [d][vo/gpu/d3d11]   yuv420p14be
@@ -219,6 +219,5 @@
 [d][vo/gpu/d3d11]   gbrapf32be
 [d][vo/gpu/d3d11]   gbrapf32 => 4 planes 1x1 32/0 [r32f/r32f/r32f/r32f] (g/b/r/a)
 [d][vo/gpu/d3d11]   opencl
-[v][vo/gpu] Testing FBO format rgba16
-[d][vo/gpu] Resizing texture: 16x16
-[v][vo/gpu] Using FBO format rgba16.
+[w][vo/gpu] High bit depth FBOs unsupported. Enabling dumb mode.
+[w][vo/gpu] Most extended features will be disabled.
```

So for me it by default picks rgba16, and is OK with that and things work. If I set rgb32f, I get the same behavior. I'll check what is checked about the FBOs, but my initial guess is that d3d11 just doesn't wörk with rgb32f and since peak calculation (or any other means of fancy) wasn't enabled, you just didn't notice. Of course this false enablement of peak calculation is incorrect, but at the very least it made you notice an issue.

For the record, locally tested rgb32f **doesn't** work and rgba32f **does** work. So it is most likely the lack or availability of the fourth plane that is the issue with d3d11.
```
[   0.079][v][vo/gpu] Testing FBO format rgba32f
[   0.079][d][vo/gpu] Resizing texture: 16x16
[   0.079][v][vo/gpu] Using FBO format rgba32f.
```

**Comment by jeeb** [comment]
I am not sure if I we can effectively error out from that function, but I did add some extra logging that should make it understandable that a user-picked FBO failed:
```
[   0.082][w][vo/gpu] User requested FBO format 'rgb32f' failed to initialize! (exists=1, renderable=0, linear_filter=0, fbo_test_result=0)
[   0.083][w][vo/gpu] High bit depth FBOs unsupported. Enabling dumb mode.
[   0.083][w][vo/gpu] Most extended features will be disabled.
```

**Comment by jeeb** [comment]
And when the FBO format does not exist at all:
```
[   0.077][w][vo/gpu] FBO format 'valpurgisnacht' not found!
[   0.077][w][vo/gpu] User requested FBO format 'valpurgisnacht' failed to initialize! (exists=0, renderable=0, linear_filter=0, fbo_test_result=0)
[   0.077][w][vo/gpu] High bit depth FBOs unsupported. Enabling dumb mode.
[   0.077][w][vo/gpu] Most extended features will be disabled.
```

**Comment by jeeb** [comment]
And I got rid of the blue screen, although you most definitely will be wanting to switch to another FBO format if you see this message:
```
[   0.079][w][vo/gpu] User requested FBO format 'rgb32f' failed to initialize! (exists=1, renderable=0, linear_filter=0, fbo_test_result=0)
[   0.079][v][vo/gpu] Disabling HDR peak computation (one or more of the following is not supported: compute shaders=1, SSBO=1, not dumb mode=0).
[   0.079][w][vo/gpu] High bit depth FBOs unsupported. Enabling dumb mode.
[   0.079][w][vo/gpu] Most extended features will be disabled.
```

**Comment by aluo-x** [comment]
Additionally here is another [log file](https://pastebin.com/k68SLgfe), but for the working [65f0825 ](https://github.com/mpv-player/mpv/commit/65f08253158ba9cb4c164a6cfd6da39b39d38e55) version.
mpv is playing the exact same file with the exact same conf.

**Comment by jeeb** [comment]
Unless you specifically disable it, it will get auto-enabled if the content needs. My fixes in the PR will disable it if dumb mode is required or set otherwise.

And yes, working log contains:
```
[   0.238][v][vo/gpu] Testing FBO format rgba16
[   0.238][d][vo/gpu] Resizing texture: 16x16
[   0.238][v][vo/gpu] Using FBO format rgba16.
```
So it is not rgb32f there. rgb32f would have never worked as far as I can see as those checks for the FBO would have never passed and dumb mode would have just been initialized. It was just never logged, and before peak calculation was not working if you couldn't get the amount of worker groups. Now we have defined the amount of working groups as 1 if the API is not available (it is not on D3D11).

tl;dr, if rgba16 does not work for some reason, try rgba32f, that seems to actually work as an FBO format in D3D11.

**Comment by jeeb** [comment]
The PR has been LGTM'd and merged, now you should get much more warnings about FBOs that you set not being usable. Additionally, even if the rendering back-end supports compute shaders, if you don't have a usable FBO all computational shader stuff is disabled (this fixes the blue screen).

So to reiterate this issue:
* Before the improvement in the peak calculation the following would happen with rgb32f:
  1. The FGO format is not usable.
  2. Peak calculation checks were more strict, and it would always be disabled with D3D11. That's why tone mapping looked so bad before with D3D11.
  3. You got basic no-FBO features enabled.
* After the improvement with rgb32f
  1. The FGO format is not usable.
  2. Peak calculation checks were made less strict because of an improvement elsewhere that enabled peak calculation to work with D3D11 (which is a major darn quality of life improvement).
  3. You got peak calculation enabled without FBOs... Blue screen as a sign of clear error.
* After the latest changes merged with rgb32f:
  1. The FGO format is not usable. A warning happens in case this FBO format was specified by the user.
  2. If there is no FBO format available, force-disable compute shaders if the back-end (like D3D11) supports them. You get a warning about that, too.
  3. You get basic no-FBO features enabled.

In other words, when you specify an invalid FBO format (or have no feasible FBO formats available), it gives you effectively the same result as before, but at least warns you about things not being enabled because of an invalid or unusable FBO format.

**Comment by jeeb** [comment]
Also this means that even if you have dumb mode, but you have both a valid FBO format as well as compute shaders available in the back-end, you will get HDR peak calculation.
